fhir-database-utils/src/main/java/com/ibm/fhir/database/utils/model/BaseObject.java

```diff
@@ -31,28 +31,34 @@
  * BaseObject
  */
 public abstract class BaseObject implements IDatabaseObject {
-    
+
     private static final Logger logger = Logger.getLogger(BaseObject.class.getName());
-    
+
     // Used to randomize a sleep after a deadlock failure
     private static final SecureRandom random = new SecureRandom();
-    
+
     private final String schemaName;
     private final String objectName;
     private final DatabaseObjectType objectType;
-    
+
     // tag map
     private final Map<String,String> tags = new HashMap<>();
-    
+
     // the database objects we depend on
     private final Set<IDatabaseObject> dependencies = new HashSet<>();
 
     // The privileges granted to different types of user
     private final Map<String, Set<Privilege>> userPrivilegeMap = new HashMap<>();
-    
+
     // The version number of the application schema this object applies to
     private final int version;
-    
+
+    // Steps to perform before updating from a previous version of this object
+    protected final List<Migration> preSteps;
+
+    // Steps to perform after updating from a previous version of this object
+    protected final List<Migration> postSteps;
```

Comment:
I've been reading through the steps included.  I think we should add Migration to the VERSION_HISTORY tracking table and DatabaseObjectType. 

Without these steps, it'd be impossible to resume the specific migration in a deterministic way.  For instance, alter table would not work again, and fail.

Reply:
The updates should happen together in a transaction and should either fail or succeed together.
The migration steps themselves should have the logic of what versions they apply to and what versions they don't.
With that said, I'm not fundamentally opposed to adding Migrations to the VERSION_HISTORY if you think its important.